pandas/core/indexes/multi.py

```diff
@@ -1235,6 +1235,9 @@ def _set_names(self, names, level=None, validate=True):
     def _get_grouper_for_level(self, mapper, level):
         indexer = self.codes[level]
         level_index = self.levels[level]
+        # break references back to ourself so that setting the name
+        # on the output of a groupby doesn't reflect back here.
+        level_index = level_index.copy()
```

Comment:
you don't need to do this here, rather do it as an else on line 1261 (add it); as take already copies, its just when that condition is not satisfied does it need a copy.

Reply:
Fixed, I think. Added a case for Categorical grouper with unobserved categories, so I think all the cases are covered.